### PR TITLE
VM: Implement, ABS, NOT opcode

### LINKED ISSUE #222: VM: implement ABS, NOT opcode

### PR DESCRIPTION
**Pull request by dauTT**
Closes #222

### CODE REVIEW

File: pkg/vm/stack/Int.go [hunk 1]
```diff
@@ -92,3 +92,15 @@ func (i *Int) Boolean() (*Boolean, error) {
 func (i *Int) Value() *big.Int {
 	return i.val
 }
+
+// Abs returns a stack integer whose underlying value is
+// the absolute value of the original stack integer.
+func (i *Int) Abs() (*Int, error) {
+	a := i.Value()
+	b, err := NewInt(a.Abs(a))
```
Comment (decentralisedkev) [inline pkg/vm/stack/Int.go]: Could you do make this :
```
 b, err := NewInt(a)
```

Comment (decentralisedkev) [reply pkg/vm/stack/Int.go]: Doing the absolute value out of the New function

File: pkg/vm/stack/stack.go [hunk 2]
```diff
@@ -148,3 +148,13 @@ func (ras *RandomAccess) PopByteArray() (*ByteArray, error) {
 	}
 	return item.ByteArray()
 }
+
+// PopBoolean will remove the last stack item that was added
+// And cast it to an stack Boolean.
```
Comment (decentralisedkev) [inline pkg/vm/stack/stack.go]: Typo: and cast it to a Boolean

File: pkg/vm/vm_ops_maths.go [hunk 3]
```diff
@@ -94,6 +94,38 @@ func Dec(op stack.Instruction, ctx *stack.Context, istack *stack.Invocation, rst
 	return NONE, nil
 }
 
+// Abs converts the Stack Item's value to an absolute value.
```
Comment (decentralisedkev) [inline pkg/vm/vm_ops_maths.go]: Abs pops an integer off of the stack and pushes it's absolute value onto the stack.
Returns an error if the popped value is not an integer or if the absolute value cannot be taken

File: pkg/vm/vm_ops_maths_test.go [hunk 4]
```diff
@@ -117,3 +117,57 @@ func TestSubOp(t *testing.T) {
...
+	}
+
+	ctx := stack.NewContext([]byte{})
+	ctx.Estack.Push(a)
+
+	v.executeOp(stack.ABS, ctx)
+
+	// Stack should have one item
+	assert.Equal(t, 1, ctx.Estack.Len())
+	assert.Equal(t, -1, ctx.Astack.Len())
```
Comment (decentralisedkev) [inline pkg/vm/vm_ops_maths_test.go]: Let's remove this assertion: `assert.Equal(t, -1, ctx.Astack.Len())` It's not consistent with the other tests, we can refactor a function out at a later stage to test the ctx object

File: pkg/vm/vm_ops_maths_test.go [hunk 5]
```diff
@@ -117,3 +117,57 @@ func TestSubOp(t *testing.T) {
...
+	assert.Equal(t, 0, ctx.IP())
+	assert.Equal(t, 0, ctx.LenInstr())
+
+	ctx.Estack.Push(b)
+
+	v.executeOp(stack.NOT, ctx)
+
+	// Stack should have one item
+	assert.Equal(t, 1, ctx.Estack.Len())
+	assert.Equal(t, -1, ctx.Astack.Len())
```
Comment (decentralisedkev) [inline pkg/vm/vm_ops_maths_test.go]: Remove:

```
	assert.Equal(t, -1, ctx.Astack.Len())
	assert.Equal(t, 0, ctx.IP())
	assert.Equal(t, 0, ctx.LenInstr())
```

Same as above comment

File: pkg/vm/stack/Int.go (hunk 1 above)
Comment (dauTT) [reply pkg/vm/stack/Int.go]: :+1:

File: pkg/vm/stack/stack.go (hunk 2 above)
Comment (dauTT) [reply pkg/vm/stack/stack.go]: :+1:

File: pkg/vm/vm_ops_maths.go (hunk 3 above)
Comment (dauTT) [reply pkg/vm/vm_ops_maths.go]: :+1:

File: pkg/vm/vm_ops_maths_test.go (hunk 4 above)
Comment (dauTT) [reply pkg/vm/vm_ops_maths_test.go]: :+1:

File: pkg/vm/vm_ops_maths_test.go (hunk 5 above)
Comment (dauTT) [reply pkg/vm/vm_ops_maths_test.go]: :+1:

**Comment by decentralisedkev** [comment]
LGTM

**Comment by hal0x2328** [comment]
100 points dauTT
50 points decentralisedkev for review

**Comment by lllwvlvwlll** [comment]
https://neoscan.io/transaction/81db58e8e3fcae21d4d614c008935de2fcb2d53bdf05ce5b4a40e13c3a110314

**Pull request merged**